include/boost/multiprecision/complex.hpp

```diff
@@ -0,0 +1,373 @@
+            {
+                // Syntax error: unexpected character
+                is.setstate(std::ios_base::failbit);
+                return is;
+            }
+        }
+        else
+        {
+            // No parentheses, just a real number from format 1
+            is.putback(ch);
+            is >> real;
+        }
+
+        if (!is.fail())
+        {
+            z.real_ = real;
+            z.imag_ = imag;
+        }
+
+        return is;
+    }
+};
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR complex<boost::multiprecision::number<T, ET>> polar(const boost::multiprecision::number<T, ET>& rho,
+                                                                                    const boost::multiprecision::number<T, ET>& theta) noexcept
+{
+    BOOST_MP_ASSERT_MSG(rho >= 0, "Rho must be positive");
+    return {rho * cos(theta), rho * sin(theta)};
+}
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR boost::multiprecision::number<T, ET> real(const complex<boost::multiprecision::number<T, ET>>& z) noexcept
+{
+    return z.real();
+}
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR boost::multiprecision::number<T, ET> imag(const complex<boost::multiprecision::number<T, ET>>& z) noexcept
+{
+    return z.imag();
+}
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR boost::multiprecision::number<T, ET> abs(const complex<boost::multiprecision::number<T, ET>>& z) noexcept
+{
+    return hypot(z.real(), z.imag());
+}
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR boost::multiprecision::number<T, ET> arg(const complex<boost::multiprecision::number<T, ET>>& z) noexcept
+{
+    return atan2(z.imag(), z.real());
+}
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR boost::multiprecision::number<T, ET> norm(const complex<boost::multiprecision::number<T, ET>>& z) noexcept
+{
+    return z.real() * z.real() + z.imag() * z.imag();
+}
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR complex<boost::multiprecision::number<T, ET>> conj(const complex<boost::multiprecision::number<T, ET>>& z) noexcept
+{
+    return {z.real(), -z.imag()};
+}
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR complex<boost::multiprecision::number<T, ET>> proj(const complex<boost::multiprecision::number<T, ET>>& z) noexcept
+{
+    if (isinf(z.real()) || isinf(z.imag()))
+    {
+        return {std::numeric_limits<boost::multiprecision::number<T, ET>>::infinity(), copysign(boost::multiprecision::number<T, ET>{0}, z.imag())};
+    }
+
+    return z;
+}
+
+template <typename T, expression_template_option ET>
+inline BOOST_MP_CXX14_CONSTEXPR complex<boost::multiprecision::number<T, ET>> exp(const complex<boost::multiprecision::number<T, ET>>& z) noexcept
```

Comment:
@mborland you're re-inventing the wheel here?  All these complex number functions are already implemented, it's just that `number<whatever>` is _already_ a complex number type with the same interface as `std::complex` for a suitably defined `whatever`.  See https://www.boost.org/doc/libs/1_85_0/libs/multiprecision/doc/html/boost_multiprecision/tut/complex.html

Reply:
In @NAThompson's original issue he was using `cpp_bin_float_100` as the number type. The templates could be `enable_if` for the already complex types, but if we want to allow a wrapper of any floating-point multiprecision type I don't see any other way around this.

Reply:
I guess my point is that adding a `boost::multiprecision::complex` class duplicates what is already there and adds confusion (and maintenance), what if you had (caution untested code ahead!!):

```
template <typename T, boost::multiprecision::expression_template_option ET>
class complex<boost::multiprecision::number<T, ET>>
{
  using self_type = complex<boost::multiprecision::number<T, ET>>;
  using complex_type = decltype(polar(std::declval<boost::multiprecision::number<T, ET>>(), std::declval<boost::multiprecision::number<T, ET>>()));
  complex_type m_data;

  complex_type& data(){ return m_data; }
  const complex_type& data()const{ return m_data; }
public:
  // public members here..

  // and then:
  friend inline self_type operator+(const self_type& a, const self_type& b)
  { return a.data() + b.data(); }
  // etc
  friend inline self_type exp(const self_type& val) 
  { return exp(val.data()); }
  // etc
};
```

Now there's no duplication, so bugs fixed in one place are fixed everywhere.  Plus if the user has included mpc.hpp for example then mpc will be used for mpfr complex number types rather than synthesising our own type.

It's still a bind writing out all the forwarding functions, but at least they're trivial and improvements in one place propagate throughout.

Reply:
I can give that a go tomorrow. Forwarding functions will be monotonous which is fine. What is non-obvious to me is how to overload say self_type + complex type like is here: https://github.com/boostorg/multiprecision/pull/620/files#diff-db4d91a1fc843b60d38c832c49b4b5ec9b8c22f772d5c5d0a56eea563ae59a5dR50. I'll play with it

Reply:
I think you have the form of the function already there?

BTW those functions are definitely NOT noexcept in the multiprecision case: memory allocations usually are involved!